### ISSUE TITLE
VICE REU tests

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/virtualc64#809]
VICE has an [excellent set of REU tests](https://sourceforge.net/p/vice-emu/code/HEAD/tree/testprogs/). Let's see how we are doing with the latest 5.1 code on the dev branch:

| Test | Status | Remark |
| ---- | ------ | -------- |
| c64transfer |  PASS |
| badoublewrite/badoublewrite |  FAIL |
| bonzai/checkchar |  PASS |
| bonzai/reudetect |  PASS |
| bonzai/spritetiming.prg | FAIL |
| colorram/colorram | PASS |
| colorram/dmacolortest | ?? |
| cpuport/cpuport | PASS |
| detectsize/detectreu | PASS |
| floatingbus/. Almost all  FAIL |
| loglitch/loglitch | PASS |
| mirrors/mirros | Almost all FAIL |
| misc/bitfill-auto | PASS |
| misc/bitfill | ?? | 
| misc/reutests.d64 | PASS |
| misc/twoblocks-auto | FAIL |
| misc/twoblocks-ff00 | FAIL |
| misc/twoblocks-ff00-auto | FAIL |
| misc/wheels.prg | PASS |
| QuickReuTest-1.1.1/quickreu-test1 | PASS |
| QuickReuTest-1.1.1/quickreu-test1-timing | PASS |
| QuickReuTest-1.1.1/quickreu-test2 | PASS |
| QuickReuTest-1.1.1/quickreu-test2-timing | PASS |
| QuickReuTest-1.1.1/quickreu-test3 | PASS |
| QuickReuTest-1.1.1/quickreu-test3-timing | FAIL |
| QuickReuTest-1.1.1/quickreu-test4 | PASS |
| QuickReuTest-1.1.1/quickreu-test4-timing | PASS |
| QuickReuTest-1.1.1/quickreu-test5 | PASS |
| QuickReuTest-1.1.1/quickreu-test5-timing | PASS |
| QuickReuTest-1.1.1/quickreu-test6 | PASS |
| QuickReuTest-1.1.1/quickreu-test6-timing | FAIL |
| QuickReuTest-1.1.1/quickreu-test7 | FAIL |
| QuickReuTest-1.1.1/quickreu-test7-timing | PASS |
| QuickReuTest-1.1.1/quickreu-test8 | PASS |
| QuickReuTest-1.1.1/quickreu-test8-timing | PASS |
| raminitpattern | All FAIL |
| ramlink/reuverify | PASS |
| ramlink/reuverify1m | - | Unsupported size |
| ramlink/reuverify2m | PASS | with REU 1750 XL |
| ramlink/reuverify4m | - | Unsupported size |
| ramlink/reuverify8m | - | Unsupported size |
| ramlink/reuverify128k | FAIL | with REU 1700 |
| ramlink/reuverify256k | FAIL | with REU 1764 |
| ramlink/reuverify512k | PASS | with REU 1750 |
| reudetect/reudetect | FAIL  | Detection works, data test fails
| reutiming2 | ??  | Test produce images. Needs closer inspection.
| rmw-trigger/rmwtrigger-ram | PASS |
| rmw-trigger/rmwtrigger-rom | PASS |
| spriteba/spriteba | ??  | Test produces images: Needs closer inspection (press number keys!).
| xfertiming/xfertiming | All FAIL  |

Bottom line: We'are still doing pretty bad.

[More infos about the VICE test bench](https://vice-emu.pokefinder.org/wiki/Testbench)

**Comment by dirkwhoffmann** [comment]
Update: Most functional bugs should be ruled out by now. The remaining issues are all timing-related, i.e., related to the timing of the BA line. I need to find more proper information about the exact behavior...

**Comment by dirkwhoffmann** [comment]
Update: Denise's source code is a good starting point for analyzing BA timing. E.g., bus blocking for the swap transfer mode is handled here:

```c++
auto Reu::swap() -> void {
    if (!swapRead) {
        if ((vicBaLow & 3) == 1) {
            if (transferLength == reg.transferLength)
                return;
        }
        ...
```

It compiles flawlessly on macOS and has a clean enough code base to insert some printf's to serve as a virtual bus analyzer. 

By analyzing data flow on the BA line in Denise, I've managed to enhance BA timing in VirtualC64 to a reasonable extent. For example, VirtualC64 does handle the late-BA timing issue for sprite 0 now, which makes it pass VICE tests `reutiming2/a.prg` and `reutiming2/b.prg`.  `reutiming3/c.prg` also works. 

Since I have bigger fish to fry, I think I'll use the current code for v5.1 and move on to other stuff. REU is still a niche feature in my eyes.
